File: zak/core/runtime/llm_agent.py
"""
ZAK Runtime — LLMAgent base class.

The LLMAgent implements the ReAct (Reason + Act) pattern, adding a proper
LLM reasoning loop on top of ZAK's existing governance infrastructure:

    Perceive (read context) → Reason (LLM call) → Act (ToolExecutor) → Observe → repeat

Key design principles:
    - Existing PolicyEngine / AuditLogger / TenantIsolation are fully preserved
    - Every LLM tool call routes through ToolExecutor (policy + audit fires)
    - max_iterations cap prevents runaway agents
    - LLM provider / model / temperature read from DSL's reasoning.llm block
    - Agents opt-in by subclassing LLMAgent instead of BaseAgent

Usage:
    class RedTeamAgent(LLMAgent):
        @property
        def tools(self):
            return [list_assets, list_vulnerabilities, compute_risk]

        def system_prompt(self, context):
            return f"You are a red team agent for tenant {context.tenant_id}..."
"""

from __future__ import annotations

import abc
import inspect
import json
import typing
import uuid
from typing import Any, Callable, Iterator, Union

from zak.core.audit.events import AuditEventType
from zak.core.audit.logger import AuditLogger
from zak.core.runtime.agent import AgentContext, AgentResult, BaseAgent
from zak.core.tools.substrate import ToolExecutor


# ---------------------------------------------------------------------------
# Schema generation — @zak_tool → OpenAI function-call JSON schema
# ---------------------------------------------------------------------------

def _build_openai_schema(tools: list[Any]) -> list[dict[str, Any]]:
    """
    Auto-generate OpenAI function-calling schema from @zak_tool decorated functions.

    Reads the function signature to infer parameter types and required fields.
    The 'context' parameter is excluded (injected automatically by ToolExecutor).
    """
    schemas: list[dict[str, Any]] = []
    for tool_fn in tools:
        meta = getattr(tool_fn, "_zak_tool", None)
        if meta is None:
            continue

        sig = inspect.signature(tool_fn)
        props: dict[str, Any] = {}
        required: list[str] = []

        _type_map: dict[Any, str] = {
            str: "string",
            int: "integer",
            float: "number",
            bool: "boolean",
            list: "array",
            dict: "object",
        }

        for param_name, param in sig.parameters.items():
            if param_name in ("context", "self"):
                continue

            ann = param.annotation
            json_type = _type_map.get(ann, "string")

            # Check for Optional[X] (Union[X, None]) via __args__
            origin = getattr(ann, "__origin__", None)
            if origin is Union or origin is typing.Union:
                args = getattr(ann, "__args__", ())
                for a in args:
                    if a is not type(None):
                        json_type = _type_map.get(a, "string")
                        break

            props[param_name] = {
                "type": json_type,
                "description": f"{param_name} parameter",
            }
            if param.default is inspect.Parameter.empty:
                required.append(param_name)

        schemas.append(
            {
                "type": "function",
                "function": {
                    "name": meta.action_id,
                    "description": meta.description or meta.name,
                    "parameters": {
                        "type": "object",
                        "properties": props,
                        "required": required,
                    },
                },
            }
        )

    return schemas


# ---------------------------------------------------------------------------
# LLMAgent base class
# ---------------------------------------------------------------------------

class LLMAgent(BaseAgent, abc.ABC):
    """
    LLM-powered agent base class using the ReAct (Reason + Act) loop.

    Subclasses define:
        - system_prompt(context) → str    — the agent's goal and persona
        - tools (property) → list         — @zak_tool functions the agent can call

    The ReAct loop, tool schema generation, and policy routing are handled
    automatically. Existing governance infrastructure is fully preserved.

    Default ReAct loop behaviour:
        1. Build initial messages (system + user goal)
        2. Call LLM → receive reasoning + tool calls
        3. Execute tool calls via ToolExecutor (policy + audit)
        4. Append tool results to conversation
        5. Repeat until LLM says "stop" or max_iterations reached
    """

    #: Default maximum ReAct iterations (override in subclass or DSL)
    max_iterations: int = 10

    def execute(self, context: AgentContext) -> AgentResult:
        from zak.core.llm.registry import get_llm_client

        # ── LLM config from DSL reasoning.llm block ────────────────────────
        llm_cfg: dict[str, Any] = {}
        if hasattr(context.dsl, "reasoning") and context.dsl.reasoning:
            r = context.dsl.reasoning
            llm_block = getattr(r, "llm", None)
            if llm_block is not None:
                llm_cfg = (
                    llm_block
                    if isinstance(llm_block, dict)
                    else llm_block.model_dump(exclude_none=True)
                )

        provider = llm_cfg.get("provider") or None
        model = llm_cfg.get("model") or None
        temperature = float(llm_cfg.get("temperature", 0.2))
        max_iter = int(llm_cfg.get("max_iterations", self.max_iterations))
        max_tokens = int(llm_cfg.get("max_tokens", 4096))

        client = get_llm_client(provider=provider, model=model)

        # ── Build initial conversation ──────────────────────────────────────
        messages: list[dict[str, Any]] = [
            {"role": "system", "content": self.system_prompt(context)},
            {
                "role": "user",
                "content": (
                    f"Execute your security analysis goal for tenant '{context.tenant_id}'. "
                    f"Environment: {context.environment}. Trace ID: {context.trace_id}. "
                    "Use your available tools to gather data, then provide a structured summary."
                ),
            },
        ]

        tools_schema = _build_openai_schema(self.tools)

        audit = AuditLogger(
            tenant_id=context.tenant_id,
            agent_id=context.agent_id,
            trace_id=context.trace_id,
        )

        total_usage: dict[str, int] = {
            "prompt_tokens": 0,
            "completion_tokens": 0,
            "total_tokens": 0,
        }
        reasoning_trace: list[dict[str, Any]] = []

        # ── ReAct loop ──────────────────────────────────────────────────────
        for iteration in range(max_iter):
            audit.log_raw(
                AuditEventType.TOOL_CALLED,
                phase="llm_reason",
                iteration=iteration + 1,
                messages_in_context=len(messages),
            )

            response = client.chat(
                messages=messages,
                tools=tools_schema,
                temperature=temperature,
                max_tokens=max_tokens,
            )

            # Accumulate token usage
            for k in total_usage:
                total_usage[k] += response.usage.get(k, 0)

            # ── LLM decided it's done (finish_reason == "stop") ────────────
            if response.finish_reason == "stop" or not response.tool_calls:
                conclusion = response.content or "Task completed successfully."
                reasoning_trace.append(
                    {
                        "iteration": iteration + 1,
                        "type": "conclusion",
                        "content": conclusion,
                    }
                )
                audit.log_raw(
                    AuditEventType.TOOL_RESULT,
                    phase="llm_conclusion",
                    iteration=iteration + 1,
                    tokens_used=total_usage["total_tokens"],
                )
                return AgentResult.ok(
                    context,
                    output={
                        "summary": conclusion,
                        "reasoning_trace": reasoning_trace,
                        "iterations": iteration + 1,
                        "llm_usage": total_usage,
                        "provider": provider or "openai",
                        "model": model,
                    },
                )

            # ── Process tool calls ─────────────────────────────────────────
            tool_results: list[dict[str, Any]] = []

            for tool_call in response.tool_calls:
                trace_entry: dict[str, Any] = {
                    "iteration": iteration + 1,
                    "type": "tool_call",
                    "tool": tool_call.name,
                    "arguments": tool_call.arguments,
                }
                reasoning_trace.append(trace_entry)

                tool_fn = self._resolve_tool(tool_call.name)
                if tool_fn is None:
                    available = [
                        getattr(t, "_zak_tool", None).action_id
                        for t in self.tools
                        if getattr(t, "_zak_tool", None)
                    ]
                    err = {
                        "error": (
                            f"Unknown tool '{tool_call.name}'. "
                            f"Available tools: {available}. "
                            "Please use ONLY the tools listed above."
                        ),
                    }
                    trace_entry["result"] = err
                    tool_results.append(self._tool_result_msg(tool_call, err))
                    continue

                try:
                    result = ToolExecutor.call(
                        tool_fn,
                        context=context,
                        **tool_call.arguments,
                    )
                    trace_entry["result"] = result
                    tool_results.append(self._tool_result_msg(tool_call, result))
                except Exception as exc:
                    err = {"error": str(exc)}
                    trace_entry["result"] = err
                    tool_results.append(self._tool_result_msg(tool_call, err))

            # Append assistant message + all tool results to context
            messages.append(
                {
                    "role": "assistant",
                    "content": response.content,
                    "tool_calls": [
                        {
                            "id": tc.id,
                            "type": "function",
                            "function": {
                                "name": tc.name,
                                "arguments": json.dumps(tc.arguments),
                            },
                        }
                        for tc in response.tool_calls
                    ],
                }
            )
            messages.extend(tool_results)

        # ── Max iterations reached without natural conclusion ───────────────
        return AgentResult.fail(
            context,
            errors=[
                f"LLM agent reached max_iterations ({max_iter}) without a conclusion. "
                "Consider increasing max_iterations in the DSL reasoning.llm block."
            ],
        )

    def execute_stream(
        self,
        context: AgentContext,
        *,
        wait_for_approval: Callable[[str], bool] | None = None,
    ) -> Iterator[dict[str, Any]]:
        """
        Streaming variant of execute(). Yields event dicts at each ReAct step
        so callers can push them to an SSE endpoint in real time.

        Args:
            context: Agent execution context.
            wait_for_approval: Optional blocking callback used for human-in-the-loop
                approval gates. Signature: (gate_id: str) -> bool (approved).
                Called when a tool's action_id appears in
                context.dsl.boundaries.approval_gates. The generator blocks until
                the callback returns. Pass None to skip gate enforcement.

        Event shapes:
          {"type": "start",             "trace_id": str, "tenant_id": str}
          {"type": "iteration",         "iteration": int}
          {"type": "tool_call",         "iteration": int, "tool": str, "arguments": dict}
          {"type": "approval_required", "gate_id": str,  "iteration": int,
                                        "tool": str,     "arguments": dict}
          {"type": "approval_granted",  "gate_id": str,  "iteration": int, "tool": str}
          {"type": "approval_denied",   "gate_id": str,  "iteration": int, "tool": str}
          {"type": "tool_result",       "iteration": int, "tool": str,
                                        "result_preview": str, "error": bool}
          {"type": "reasoning",         "iteration": int, "content": str}
          {"type": "complete",          "output": dict,  "iterations": int, "llm_usage": dict}
          {"type": "error",             "message": str}
        """
        from zak.core.llm.registry import get_llm_client

        # Build the approval-gate set from DSL boundaries once
        _approval_gates: set[str] = set()
        if hasattr(context.dsl, "boundaries") and context.dsl.boundaries:
            _approval_gates = set(context.dsl.boundaries.approval_gates or [])

        yield {"type": "start", "trace_id": context.trace_id, "tenant_id": context.tenant_id}

        # ── LLM config from DSL ────────────────────────────────────────────
        llm_cfg: dict[str, Any] = {}
        if hasattr(context.dsl, "reasoning") and context.dsl.reasoning:
            r = context.dsl.reasoning
            llm_block = getattr(r, "llm", None)
            if llm_block is not None:
                llm_cfg = (
                    llm_block
                    if isinstance(llm_block, dict)
                    else llm_block.model_dump(exclude_none=True)
                )

        provider = llm_cfg.get("provider") or None
        model = llm_cfg.get("model") or None
        temperature = float(llm_cfg.get("temperature", 0.2))
        max_iter = int(llm_cfg.get("max_iterations", self.max_iterations))
        max_tokens = int(llm_cfg.get("max_tokens", 4096))

        try:
            client = get_llm_client(provider=provider, model=model)
        except Exception as exc:
            yield {"type": "error", "message": f"LLM client init failed: {exc}"}
            return

        # ── Build initial conversation ─────────────────────────────────────
        messages: list[dict[str, Any]] = [
            {"role": "system", "content": self.system_prompt(context)},
            {
                "role": "user",
                "content": (
                    f"Execute your security analysis goal for tenant '{context.tenant_id}'. "
                    f"Environment: {context.environment}. Trace ID: {context.trace_id}. "
                    "Use your available tools to gather data, then provide a structured summary."
                ),
            },
        ]

        tools_schema = _build_openai_schema(self.tools)

        audit = AuditLogger(
            tenant_id=context.tenant_id,
            agent_id=context.agent_id,
            trace_id=context.trace_id,
        )

        total_usage: dict[str, int] = {
            "prompt_tokens": 0,
            "completion_tokens": 0,
            "total_tokens": 0,
        }
        reasoning_trace: list[dict[str, Any]] = []

        # ── ReAct loop ─────────────────────────────────────────────────────
        for iteration in range(max_iter):
            yield {"type": "iteration", "iteration": iteration + 1}

            audit.log_raw(
                AuditEventType.TOOL_CALLED,
                phase="llm_reason",
                iteration=iteration + 1,
                messages_in_context=len(messages),
            )

            try:
                response = client.chat(
                    messages=messages,
                    tools=tools_schema,
                    temperature=temperature,
                    max_tokens=max_tokens,
                )
            except Exception as exc:
                yield {
                    "type": "error",
                    "message": f"LLM call failed at iteration {iteration + 1}: {exc}",
                }
                return

            for k in total_usage:
                total_usage[k] += response.usage.get(k, 0)

            # ── LLM decided it's done ──────────────────────────────────────
            if response.finish_reason == "stop" or not response.tool_calls:
                conclusion = response.content or "Task completed successfully."
                reasoning_trace.append(
                    {"iteration": iteration + 1, "type": "conclusion", "content": conclusion}
                )
                yield {
                    "type": "reasoning",
                    "iteration": iteration + 1,
                    "content": conclusion[:1000],
                }
                audit.log_raw(
                    AuditEventType.TOOL_RESULT,
                    phase="llm_conclusion",
                    iteration=iteration + 1,
                    tokens_used=total_usage["total_tokens"],
                )
                output = {
                    "summary": conclusion,
                    "reasoning_trace": reasoning_trace,
                    "iterations": iteration + 1,
                    "llm_usage": total_usage,
                    "provider": provider or "openai",
                    "model": model,
                }
                yield {
                    "type": "complete",
                    "output": output,
                    "iterations": iteration + 1,
                    "llm_usage": total_usage,
                }
                return

            # ── Process tool calls ─────────────────────────────────────────
            tool_results: list[dict[str, Any]] = []

            for tool_call in response.tool_calls:
                yield {
                    "type": "tool_call",
                    "iteration": iteration + 1,
                    "tool": tool_call.name,
                    "arguments": tool_call.arguments,
                }

                trace_entry: dict[str, Any] = {
                    "iteration": iteration + 1,
                    "type": "tool_call",
                    "tool": tool_call.name,
                    "arguments": tool_call.arguments,
                }
                reasoning_trace.append(trace_entry)

                # ── Approval gate check ────────────────────────────────────
                if tool_call.name in _approval_gates and wait_for_approval is not None:
                    gate_id = uuid.uuid4().hex[:12]
                    yield {
                        "type": "approval_required",
                        "gate_id": gate_id,
                        "iteration": iteration + 1,
                        "tool": tool_call.name,
                        "arguments": tool_call.arguments,
                    }
                    approved = wait_for_approval(gate_id)
                    if approved:
                        yield {
                            "type": "approval_granted",
                            "gate_id": gate_id,
                            "iteration": iteration + 1,
                            "tool": tool_call.name,
                        }
                    else:
                        denied_msg = f"Tool '{tool_call.name}' denied by human reviewer."
                        yield {
                            "type": "approval_denied",
                            "gate_id": gate_id,
                            "iteration": iteration + 1,
                            "tool": tool_call.name,
                        }
                        denied_result = {"error": denied_msg}
                        trace_entry["result"] = denied_result
                        tool_results.append(self._tool_result_msg(tool_call, denied_result))
                        continue  # skip tool execution; LLM sees the denial in next iteration
                # ── /Approval gate check ───────────────────────────────────

                tool_fn = self._resolve_tool(tool_call.name)
                if tool_fn is None:
                    available = [
                        getattr(t, "_zak_tool", None).action_id
                        for t in self.tools
                        if getattr(t, "_zak_tool", None)
                    ]
                    err = {
                        "error": (
                            f"Unknown tool '{tool_call.name}'. "
                            f"Available tools: {available}. "
                            "Please use ONLY the tools listed above."
                        ),
                    }
                    trace_entry["result"] = err
                    tool_results.append(self._tool_result_msg(tool_call, err))
                    yield {
                        "type": "tool_result",
                        "iteration": iteration + 1,
                        "tool": tool_call.name,
                        "result_preview": str(err)[:300],
                        "error": True,
                    }
                    continue

                try:
                    result = ToolExecutor.call(
                        tool_fn,
                        context=context,
                        **tool_call.arguments,
                    )
                    trace_entry["result"] = result
                    tool_results.append(self._tool_result_msg(tool_call, result))
                    yield {
                        "type": "tool_result",
                        "iteration": iteration + 1,
                        "tool": tool_call.name,
                        "result_preview": str(result)[:300],
                        "error": False,
                    }
                except Exception as exc:
                    err = {"error": str(exc)}
                    trace_entry["result"] = err
                    tool_results.append(self._tool_result_msg(tool_call, err))
                    yield {
                        "type": "tool_result",
                        "iteration": iteration + 1,
                        "tool": tool_call.name,
                        "result_preview": str(err)[:300],
                        "error": True,
                    }

            # Append assistant + tool results to context
            messages.append(
                {
                    "role": "assistant",
                    "content": response.content,
                    "tool_calls": [
                        {
                            "id": tc.id,
                            "type": "function",
                            "function": {
                                "name": tc.name,
                                "arguments": json.dumps(tc.arguments),
                            },
                        }
                        for tc in response.tool_calls
                    ],
                }
            )
            messages.extend(tool_results)

        # ── Max iterations reached ─────────────────────────────────────────
        yield {
            "type": "error",
            "message": (
                f"LLM agent reached max_iterations ({max_iter}) without a conclusion. "
                "Consider increasing max_iterations in the DSL reasoning.llm block."
            ),
        }

    # ── Helpers ────────────────────────────────────────────────────────────

    def _resolve_tool(self, action_id: str) -> Any | None:
        """Find the @zak_tool function by its action_id from this agent's tool list."""
        for fn in self.tools:
            meta = getattr(fn, "_zak_tool", None)
            if meta and meta.action_id == action_id:
                return fn
        return None

    def _tool_result_msg(self, tool_call: Any, result: Any) -> dict[str, Any]:
        """Format a tool result as an OpenAI-compatible tool message."""
        return {
            "role": "tool",
            "tool_call_id": tool_call.id,
            "content": (
                json.dumps(result)
                if not isinstance(result, str)
                else result
            ),
        }

    # ── Abstract interface ─────────────────────────────────────────────────

    @abc.abstractmethod
    def system_prompt(self, context: AgentContext) -> str:
        """
        Define the agent's goal and persona.

        This is the system prompt sent to the LLM. It should:
        - State the agent's security objective clearly
        - Describe the expected tool call sequence
        - Specify the desired output format (structured JSON)
        - Remind the LLM to base every claim on tool output

        Example:
            return f"You are a risk quantification agent for tenant '{context.tenant_id}'.
            Goal: Compute risk scores for all assets..."
        """

    @property
    @abc.abstractmethod
    def tools(self) -> list[Any]:
        """
        Return the list of @zak_tool functions this agent may call.

        These functions must also be declared in the agent's DSL
        capabilities.tools list (otherwise ToolExecutor will reject them).

        Example:
            return [list_assets, list_vulnerabilities, compute_risk, write_risk_node]
        """

File: adk/go/pkg/runtime/llm_agent.go
package runtime

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/securezeron/zeron-agent-development-kit/adk/go/pkg/audit"
	"github.com/securezeron/zeron-agent-development-kit/adk/go/pkg/llm"
)

// ---------------------------------------------------------------------------
// LLMAgent interface
// ---------------------------------------------------------------------------

// LLMAgent extends BaseAgent with LLM-powered ReAct (Reason + Act) support.
//
// Implementations provide:
//   - SystemPrompt(ctx) — the agent's goal and persona as a system message
//   - Tools()           — the list of ZakTool pointers the agent may call
//
// The ReAct loop, tool schema generation, and tool execution are handled
// automatically by RunLLMAgent.
type LLMAgent interface {
	// SystemPrompt returns the system prompt that defines the agent's persona,
	// goal, and expected output format for the given execution context.
	SystemPrompt(ctx *AgentContext) string

	// Tools returns the list of tools this agent is allowed to invoke.
	// Each tool carries metadata and a function pointer.
	Tools() []*LLMAgentTool
}

// LLMAgentTool is a lightweight tool descriptor used by the LLM agent.
// It mirrors the fields needed for OpenAI function-call schema generation
// and tool dispatch without importing the tools package (to avoid cycles).
type LLMAgentTool struct {
	// ActionID is the unique identifier used in function-call schemas.
	ActionID string
	// Name is the human-readable tool name.
	Name string
	// Description is a short sentence describing what the tool does.
	Description string
	// Parameters describes the JSON Schema properties for the tool.
	// Keys are parameter names; values are maps with "type" and "description".
	Parameters []LLMToolParam
	// Fn is the function to execute when the tool is called.
	Fn func(ctx *AgentContext, args map[string]interface{}) (interface{}, error)
}

// LLMToolParam describes a single parameter for an LLMAgentTool.
type LLMToolParam struct {
	Name        string `json:"name"`
	Type        string `json:"type"` // "string", "integer", "number", "boolean", "array", "object"
	Description string `json:"description"`
	Required    bool   `json:"required"`
}

// ---------------------------------------------------------------------------
// BuildOpenAISchema — convert tools to OpenAI function-call JSON
// ---------------------------------------------------------------------------

// BuildOpenAISchema converts a list of LLMAgentTool pointers into the
// OpenAI function-calling schema format (list of {"type":"function",...}).
func BuildOpenAISchema(tools []*LLMAgentTool) []map[string]interface{} {
	schemas := make([]map[string]interface{}, 0, len(tools))
	for _, tool := range tools {
		props := make(map[string]interface{})
		required := make([]string, 0)

		for _, p := range tool.Parameters {
			props[p.Name] = map[string]interface{}{
				"type":        p.Type,
				"description": p.Description,
			}
			if p.Required {
				required = append(required, p.Name)
			}
		}

		schemas = append(schemas, map[string]interface{}{
			"type": "function",
			"function": map[string]interface{}{
				"name":        tool.ActionID,
				"description": tool.Description,
				"parameters": map[string]interface{}{
					"type":       "object",
					"properties": props,
					"required":   required,
				},
			},
		})
	}
	return schemas
}

// ---------------------------------------------------------------------------
// RunLLMAgent — the ReAct loop
// ---------------------------------------------------------------------------

// RunLLMAgent executes an LLMAgent using the ReAct (Reason + Act) loop.
//
// Loop:
//  1. Build initial messages (system + user goal).
//  2. Call LLM -> receive reasoning + tool calls.
//  3. Execute tool calls (via the tool's Fn).
//  4. Append tool results to conversation.
//  5. Repeat until LLM says "stop" or max_iterations reached.
//
// The LLM client, temperature, max_tokens, and max_iterations are read from
// the DSL's reasoning.llm block. If no LLM config is present, defaults are
// used (provider=mock, temperature=0.2, max_tokens=4096, max_iterations=10).
func RunLLMAgent(agent LLMAgent, ctx *AgentContext) *AgentResult {
	start := time.Now()

	// ── Resolve LLM config from DSL ──────────────────────────────────────
	provider := ""
	model := ""
	temperature := 0.2
	maxIter := 10
	maxTokens := 4096

	if ctx.DSL != nil && ctx.DSL.Reasoning.LLM != nil {
		llmCfg := ctx.DSL.Reasoning.LLM
		if llmCfg.Provider != "" {
			provider = llmCfg.Provider
		}
		if llmCfg.Model != "" {
			model = llmCfg.Model
		}
		if llmCfg.Temperature > 0 {
			temperature = llmCfg.Temperature
		}
		if llmCfg.MaxIterations > 0 {
			maxIter = llmCfg.MaxIterations
		}
		if llmCfg.MaxTokens > 0 {
			maxTokens = llmCfg.MaxTokens
		}
	}

	// ── Get LLM client ───────────────────────────────────────────────────
	client, err := llm.GetLLMClient(provider, model, "", "")
	if err != nil {
		durationMs := float64(time.Since(start).Milliseconds())
		return ResultFail(ctx, []string{fmt.Sprintf("LLM client init failed: %s", err)}, durationMs)
	}

	// ── Build initial conversation ───────────────────────────────────────
	messages := []map[string]interface{}{
		{"role": "system", "content": agent.SystemPrompt(ctx)},
		{
			"role": "user",
			"content": fmt.Sprintf(
				"Execute your security analysis goal for tenant '%s'. "+
					"Environment: %s. Trace ID: %s. "+
					"Use your available tools to gather data, then provide a structured summary.",
				ctx.TenantID, ctx.Environment, ctx.TraceID,
			),
		},
	}

	toolsSchema := BuildOpenAISchema(agent.Tools())

	// Build a lookup map for tool dispatch.
	toolMap := make(map[string]*LLMAgentTool)
	for _, t := range agent.Tools() {
		toolMap[t.ActionID] = t
	}

	logger := audit.NewLogger(ctx.TenantID, ctx.AgentID(), ctx.TraceID)

	totalUsage := map[string]int{
		"prompt_tokens":     0,
		"completion_tokens": 0,
		"total_tokens":      0,
	}
	reasoningTrace := make([]map[string]interface{}, 0)

	// ── ReAct loop ───────────────────────────────────────────────────────
	for iteration := 0; iteration < maxIter; iteration++ {
		logger.LogRaw(audit.ToolCalled, map[string]interface{}{
			"phase":              "llm_reason",
			"iteration":          iteration + 1,
			"messages_in_context": len(messages),
		})

		response, err := client.Chat(messages, toolsSchema, maxTokens, temperature)
		if err != nil {
			durationMs := float64(time.Since(start).Milliseconds())
			return ResultFail(ctx, []string{
				fmt.Sprintf("LLM call failed at iteration %d: %s", iteration+1, err),
			}, durationMs)
		}

		// Accumulate token usage.
		for k := range totalUsage {
			totalUsage[k] += response.Usage[k]
		}

		// ── LLM decided it is done (finish_reason == "stop") ─────────
		if response.FinishReason == "stop" || len(response.ToolCalls) == 0 {
			conclusion := "Task completed successfully."
			if response.Content != nil && *response.Content != "" {
				conclusion = *response.Content
			}
			reasoningTrace = append(reasoningTrace, map[string]interface{}{
				"iteration": iteration + 1,
				"type":      "conclusion",
				"content":   conclusion,
			})

			logger.LogRaw(audit.ToolResult, map[string]interface{}{
				"phase":       "llm_conclusion",
				"iteration":   iteration + 1,
				"tokens_used": totalUsage["total_tokens"],
			})

			durationMs := float64(time.Since(start).Milliseconds())
			return ResultOk(ctx, map[string]interface{}{
				"summary":         conclusion,
				"reasoning_trace": reasoningTrace,
				"iterations":      iteration + 1,
				"llm_usage":       totalUsage,
				"provider":        provider,
				"model":           model,
			}, durationMs)
		}

		// ── Process tool calls ───────────────────────────────────────
		toolResults := make([]map[string]interface{}, 0)

		for _, tc := range response.ToolCalls {
			traceEntry := map[string]interface{}{
				"iteration": iteration + 1,
				"type":      "tool_call",
				"tool":      tc.Name,
				"arguments": tc.Arguments,
			}
			reasoningTrace = append(reasoningTrace, traceEntry)

			toolDef, found := toolMap[tc.Name]
			if !found {
				errResult := map[string]interface{}{"error": fmt.Sprintf("Unknown tool: %s", tc.Name)}
				traceEntry["result"] = errResult
				toolResults = append(toolResults, toolResultMsg(tc, errResult))
				continue
			}

			result, execErr := toolDef.Fn(ctx, tc.Arguments)
			if execErr != nil {
				errResult := map[string]interface{}{"error": execErr.Error()}
				traceEntry["result"] = errResult
				toolResults = append(toolResults, toolResultMsg(tc, errResult))
			} else {
				traceEntry["result"] = result
				toolResults = append(toolResults, toolResultMsg(tc, result))
			}
		}

		// Append assistant message + all tool results to context.
		assistantToolCalls := make([]interface{}, 0, len(response.ToolCalls))
		for _, tc := range response.ToolCalls {
			argsJSON, _ := json.Marshal(tc.Arguments)
			assistantToolCalls = append(assistantToolCalls, map[string]interface{}{
				"id":   tc.ID,
				"type": "function",
				"function": map[string]interface{}{
					"name":      tc.Name,
					"arguments": string(argsJSON),
				},
			})
		}

		contentVal := interface{}(nil)
		if response.Content != nil {
			contentVal = *response.Content
		}
		messages = append(messages, map[string]interface{}{
			"role":       "assistant",
			"content":    contentVal,
			"tool_calls": assistantToolCalls,
		})
		for _, tr := range toolResults {
			messages = append(messages, tr)
		}
	}

	// ── Max iterations reached ───────────────────────────────────────────
	durationMs := float64(time.Since(start).Milliseconds())
	return ResultFail(ctx, []string{
		fmt.Sprintf(
			"LLM agent reached max_iterations (%d) without a conclusion. "+
				"Consider increasing max_iterations in the DSL reasoning.llm block.",
			maxIter,
		),
	}, durationMs)
}

// toolResultMsg formats a tool result as an OpenAI-compatible tool message.
func toolResultMsg(tc llm.ToolCall, result interface{}) map[string]interface{} {
	var content string
	switch v := result.(type) {
	case string:
		content = v
	default:
		b, err := json.Marshal(result)
		if err != nil {
			content = fmt.Sprintf("%v", result)
		} else {
			content = string(b)
		}
	}

	return map[string]interface{}{
		"role":         "tool",
		"tool_call_id": tc.ID,
		"content":      content,
	}
}
